### ISSUE TITLE
[Clang] false-positive of `-Wweak-vtables` with `[[clang::exclude_from_explicit_instantiation]]`

### ISSUE THREAD

**Issue by philnik777** [llvm/llvm-project#195110]
```c++
template <class> struct basic_streambuf {
  __attribute__((__exclude_from_explicit_instantiation__)) virtual void
  overflow();

  virtual ~basic_streambuf() {}
};
extern template class basic_streambuf<char>; // warning: 'basic_streambuf<char>' has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit

basic_streambuf<char> b;
```
Which is obviously bogus. This bug has been introduced recently (it's not in any release yet) and causes problems for libc++, since it makes the warning unusable for us. It would be great if this could be fixed soon.

CC @AaronBallman

**Comment by AaronBallman** [comment]
Possibly related to changes in https://github.com/llvm/llvm-project/pull/191658 but I've not done a bisect

CC @mizvekov @16bit-ykiko @erichkeane @zyn0217 in case this rings a bell for any of you

**Comment by 16bit-ykiko** [comment]
> Possibly related to changes in [#191658](https://github.com/llvm/llvm-project/pull/191658) but I've not done a bisect
> 
> CC [@mizvekov](https://github.com/mizvekov) [@16bit-ykiko](https://github.com/16bit-ykiko) [@erichkeane](https://github.com/erichkeane) [@zyn0217](https://github.com/zyn0217) in case this rings a bell for any of you

It seems that this is not caused by #191658. The regression was introduced by https://github.com/llvm/llvm-project/pull/168171 (commit 9ce9e4e6c19be).

Before that PR, the code in `Sema::DefineUsedVTables` unconditionally suppressed vtable emission for explicit instantiation declarations:

https://github.com/llvm/llvm-project/blob/2205a23801d4e0cca3cd28670b611a7f6adbeb06/clang/lib/Sema/SemaDeclCXX.cpp#L19179-L19188

This meant the code would `continue` early and **never reach the `-Wweak-vtables` warning check**.

After that PR, the logic was changed to keep `DefineVTable = true` when any method has `exclude_from_explicit_instantiation`:

https://github.com/llvm/llvm-project/blob/afdbe7bc4924a8b263c87e3fab4cab213fcaa9ff/clang/lib/Sema/SemaDeclCXX.cpp#L19252-L19265

Now the code falls through to the warning check, which doesn't filter out `TSK_ExplicitInstantiationDeclaration`:

https://github.com/llvm/llvm-project/blob/afdbe7bc4924a8b263c87e3fab4cab213fcaa9ff/clang/lib/Sema/SemaDeclCXX.cpp#L19288-L19297

The warning fires because `TSK_ExplicitInstantiationDeclaration` is not excluded here.

**Possible fix:** Add `TSK_ExplicitInstantiationDeclaration` to the filter:

```cpp
    ClassTSK != TSK_ImplicitInstantiation &&
    ClassTSK != TSK_ExplicitInstantiationDefinition &&
    ClassTSK != TSK_ExplicitInstantiationDeclaration) {
```

The warning is bogus because the vtable is still provided by the explicit instantiation definition TU — `exclude_from_explicit_instantiation` only excludes individual method bodies, not the vtable itself.

https://github.com/llvm/llvm-project/pull/168171 (commit 9ce9e4e6c19be) haven't been cherry picked to llvm 22.1.4, so you won't find error in the recent releases. 

CC @kikairoya @zmodem, any ideas?

**Comment by AaronBallman** [comment]
> It seems that this is not caused by https://github.com/llvm/llvm-project/pull/191658. The regression was introduced by https://github.com/llvm/llvm-project/pull/168171 (commit https://github.com/llvm/llvm-project/commit/9ce9e4e6c19bed1781db2c20fae7f24c905438e9).

Thank you for the investigation, I appreciate it!

**Comment by kikairoya** [comment]
Thank you for the report and analysis.
I've submitted a fix as suggested: #195189

I'm heading to work now and will be unavailable, so please feel free to modify the PR if necessary.
